File: zig/luhn/luhn.zig
const std = @import("std");

pub fn isValid(comptime s: []const u8) bool {
    var num: [s.len]u8 = undefined;
    var cursor: usize = 0;
    for (s) |c| {
        num[cursor] = std.fmt.charToDigit(c, 10) catch if (c == ' ') continue else return false;
        cursor += 1;
    }

    var j: usize = 1;
    while (j < cursor) : (j += 2) {
        const double = num[cursor - j - 1] * 2;
        num[cursor - j - 1] = if (double > 9) double - 9 else double;
    }

    var sum: u64 = 0;
    for (num) |n| {
        sum += n;
    }
    return cursor > 1 and sum % 10 == 0;
}

File: c/luhn/luhn.c
#include "luhn.h"

#include <ctype.h>

bool luhn(const char *num) {
  int digits[128];
  int cursor = 0;
  for (int i = 0; num[i] != '\0'; ++i) {
    if (isdigit(num[i])) {
      digits[cursor++] = num[i] - '0';
    } else if (num[i] != ' ') {
      return false;
    }
  }

  int i = 1;
  while (i < cursor) {
    const int doubled = digits[cursor - i - 1] * 2;
    digits[cursor - i - 1] = doubled > 9 ? doubled - 9 : doubled;
    i += 2;
  }

  int sum = 0;
  for (int i = 0; i < cursor; ++i) {
    sum += digits[i];
  }
  return cursor > 1 && sum % 10 == 0;
}
